### PR TITLE
Correct rank-1 matmul for signed strides

### LINKED ISSUE #1172: Add NumPy-compatible batched matmul with layout-aware HPC execution
## Motivation

`SimpleArray::matmul()` accepts rank-1 and rank-2 operands but rejects leading batch dimensions. This prevents NumPy-style batched matrix multiplication:

```python
lhs.shape == (2, 1, 3, 4)
rhs.shape == (1, 5, 4, 6)

result = lhs @ rhs
assert result.shape == (2, 5, 3, 6)
```

Only the leading batch axes should broadcast. The trailing axes retain their vector and matrix roles, and the contracted `K` dimensions must match. Existing rank-1 and rank-2 result conventions remain unchanged.

The implementation must also be layout-aware. A loop over broadcast coordinates is correct but can repeatedly copy the same non-contiguous matrix before calling BLAS. On Apple M1, NumPy and the prototype both use Accelerate. For `(1,256,256) @ (64,256,256)`, the dense route is at parity, while a broadcast negative-stride lhs exposes the cost of repeated packing:

| Lhs supplied to the call | Lhs element strides | NumPy | Prototype |
| --- | --- | ---: | ---: |
| Dense | `(65536,256,1)` | 11.960 ms | 11.873 ms |
| Reversed columns | `(65536,256,-1)` | 1997.845 ms | 12.099 ms |
| Same reversed values, prepacked before timing | `(65536,256,1)` | 11.967 ms | 11.950 ms |

The prepacked control shows that broadcasting itself is not the slow operation. The prototype packs the one supplied lhs inside the timed call and reuses it for 64 GEMMs. Full platform results and controls are kept in the linked fork study.

The prototype also found an existing signed-stride correctness defect that must be fixed before optimized routes are reused:

```python
values = np.arange(8.0)[::-1]
expected = np.matmul(values, values)  # 140.0

array = solvcon.SimpleArrayFloat64(array=values)
actual = array.matmul(array)           # 56.0 in the current path
```

## Proposed design

Callers continue to use `lhs.matmul(rhs)` or `lhs @ rhs`. Planning and execution remain internal:

```text
SimpleArray::matmul(rhs)
        |
        v
MatmulPlan::make(lhs, rhs)
        |
        +-- validate vector roles and contracted K
        +-- compute broadcast batch and output shapes
        `-- retain signed matrix and batch strides
        |
        v
MatmulExecutor::execute(plan, lhs, rhs)
        +-- generic signed-stride contraction
        +-- direct DOT, GEMV, or GEMM
        `-- pack each unsupported supplied operand once, then use BLAS
```

The batch-coordinate layer stores only an iteration shape and operand offset mappings. It does not contain dtype policy, kernels, BLAS decisions, or Python objects. `MatmulPlan` owns matrix semantics, while `MatmulExecutor` owns backend and packing decisions.

Broadcasting is represented by zero batch strides rather than expanded arrays. For `(2,1,M,K) @ (1,5,K,N)`, execution performs ten contractions. If both supplied operands need packing, it copies the two lhs matrices and five rhs matrices, rather than copying one lhs and one rhs for each of the ten result coordinates.

A newly allocated result is C-contiguous, so each output matrix can be a direct BLAS destination. Non-contiguous destination support is outside this change.

The shared batch-coordinate types will be introduced together with their first matmul consumer. This issue does not require a universal operation object or unused elementwise and reduction infrastructure.

## Migration path

Tasks 2 through 9 develop and validate the new implementation through a temporary `matmul_planned()` entry point. Existing `matmul()`, `matmul_fast()`, and `matmul_blas()` remain as correctness and performance controls during development.

Task 10 makes `matmul()` and `operator@` use the planned implementation, removes the legacy entry points, and removes the temporary staging API.

## Implementation outline

Keep each pull request near 200 to 300 changed lines, including implementation, tests, and profiling changes. Every new abstraction must be consumed by matmul in the same task that introduces it.

Dependencies: Tasks 1 and 2 may proceed in parallel. Task 3 follows Task 2; Task 4 follows Task 3; Task 5 follows Task 4; Task 6 requires Tasks 1 and 5; Task 7 follows Task 6; Task 8 follows Task 7; Task 9 follows Task 8; Task 10 requires Tasks 7 through 9.

- [x] **Task 1: Correct rank-1 signed strides.** Use strided element access for both vectors and add Python differential tests for contiguous, negative, and step-two layouts. (Resolved by #1173 )
- [x] **Task 2: Establish rank-2 planned execution.** Add `matmul_planned()`, a matrix-only `MatmulPlan`, a generic signed-stride executor, differential tests, and a shared-profiler entry. (Resolved by #1174 )
- [x] **Task 3: Traverse equal batch shapes.** Add the batch loop and operand mappings with an immediate `(B,M,K) @ (B,K,N)` generic consumer, tests, and profiles. (Resolved by #1181 )
- [x] **Task 4: Broadcast batch axes.** Rank-align leading axes with zero strides and test `(2,1,M,K) @ (1,5,K,N)` across supported layouts. (Resolved by #1208 )
- [x] **Task 5: Complete generic operand roles.** Add vector-vector, vector-matrix, matrix-vector, and batched role and result-shape rules with differential tests and profiles. (Resolved by #1208 )
- [x] **Task 6: Add direct BLAS dispatch.** Route compatible DOT, GEMV, and GEMM layouts, including C- and F-contiguous matrices, to BLAS without copying and measure through at least `1024x1024`. (Resolved by #1209 )
- [x] **Task 7: Pack unsupported matrices once.** Pack each reusable supplied matrix before batch traversal and validate copy and GEMM counts on OpenBLAS and Accelerate. (Resolved by #1256)
- [x] **Task 8: Tune batched vector routes and expose execution decisions.** Represent the selected route and packing policy explicitly, then add direct or pack-once GEMV only where OpenBLAS and Accelerate measurements beat generic traversal. (Resolved by #1259)
- [x] **Task 9: Tune small reused matrix routes.** Add an out-of-line static IKJ candidate for measured small square GEMM shapes where packing does not amortize, with correctness and platform profiles. (Resolved by #1283)
- [x] **Task 10: Replace the legacy APIs.** Move `matmul()` and `operator@` to planned execution, remove legacy and staging entry points, and rerun the full tests and profiles. (#1354)

Generated policies, forced-route tuning, and build-time empirical search are follow-up work outside this issue. The execution decision introduced in Task 8 is the seam they will consume.

## Global acceptance

- The public operation and forced execution routes match NumPy values for all supported shapes and layouts, subject to existing `SimpleArray` result-shape conventions.
- Timed calls include planning, allocation, packing, dispatch, and arithmetic, with thread configuration applied before importing NumPy.
- Batch broadcasting uses zero strides and does not expand operands.
- The final public API contains no temporary `matmul_planned()`, public `matmul_fast()`, or public `matmul_blas()` entry point.

## References

- [Matmul dispatch and broadcasting validation: design, methodology, and reproduction guide](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/matmul-validation.md)
- [Prototype class boundaries and rationale](https://github.com/ThreeMonth03/solvcon/issues/22#issuecomment-5071073901)
- [NumPy source walkthrough for repeated input packing and output copy-back](https://github.com/ThreeMonth03/solvcon/issues/22#issuecomment-5068663861)

## Appendix: prototype and benchmark evidence

- [Prototype draft PR #26](https://github.com/ThreeMonth03/solvcon/pull/26)
- Cross-platform topology results: [Ubuntu](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/ubuntu-matmul-topology-results.md) and [Apple Silicon](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/macos-matmul-topology-results.md)
- Broadcast scaling and route-count experiments: [Ubuntu](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/ubuntu-matmul-broadcast-scaling-results.md) and [Apple Silicon](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/macos-matmul-broadcast-scaling-results.md)
- Exhaustive 31,825 layout-pair validation: [Apple Silicon summary](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/macos-matmul-cartesian-summary.md) and [reproduction script](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/profiling/profile_matmul_cartesian.py)
- Pack-once crossover study: [Ubuntu results](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/doc/source/devplan/layout-aware-execution/ubuntu-matmul-pack-crossover-results.md) and [reproduction script](https://github.com/ThreeMonth03/solvcon/blob/codex/prototype-layout-aware-execution/profiling/profile_matmul_pack_crossover.py)

### PR DESCRIPTION
**Pull request by ThreeMonth03**
Rank-1 matmul currently uses strided indexing for the lhs but raw contiguous buffer indexing for the rhs. A reversed rhs therefore returns an incorrect dot product:

```python
values = np.arange(8.0)[::-1]
array = solvcon.SimpleArrayFloat64(array=values)

np.matmul(values, values)  # 140.0
array.matmul(array)        # 56.0 before this change
```

This change reads the rhs through the existing `operator()` strided accessor, matching the lhs and the other generic matmul paths. It introduces no new accessor or execution abstraction.

Differential unit tests cover the full Cartesian product of contiguous, negative, and step-two lhs/rhs layouts for float32 and float64. The same cases run through `matmul()`, `matmul_fast()`, and `matmul_blas()`.

Related to #1172.

### CODE REVIEW

File: cpp/solvcon/buffer/matmul.hpp [hunk 1]
```diff
@@ -282,7 +282,7 @@ A SimpleArrayMatmulHelper<A, T>::matmul_vec_vec()
     value_type v = 0;
     for (ssize_t i = 0; i < k; ++i)
     {
-        v += m_lhs(i) * m_rhs.data(i);
+        v += m_lhs(i) * m_rhs(i);
```
Comment (ThreeMonth03) [inline cpp/solvcon/buffer/matmul.hpp:449]: `SimpleArray::data(ssize_t i)` calculates offset without considering strides, so I replace `SimpleArray::data(ssize_t i)` with operator, which considers strides when calculating data offset.

**Review by ThreeMonth03 — COMMENTED** [review]
@yungyuc Please take a look.

**Review by yungyuc — APPROVED** [review]
LGTM
